cors/304.htm

```diff
@@ -0,0 +1,105 @@
+  var client = new XMLHttpRequest()
+  client.error_caught = false
+  try {
+    client.open('GET', url, true)
+    client.setRequestHeader(state_header, id)
+    client.send()
+  } catch (e) {
+    client.error_caught = e
+  }
+  client.onreadystatechange = function() {
+    if (client.readyState == client.DONE) {
+      console.log('finished request ' + id)
+      t.step(function() {
+        assert_true(client.status != 299, "req " + id + " server says: " + client.responseText)
+      })
+      ready(client)
+    }
+  }
+  return client
+}
+
+function two_reqs(id1, id2, should_have_same_body, t, done) {
+  var rand = Date.now()
+  var url = CROSSDOMAIN + 'resources/304.py?id=' + id1 + '&r=%s' + rand
+
+  var client1 = req(url, id1, t, function(client1) {
+    t.step(function() {
+      assert_equals(client1.response, "Success", "didn't get successful 1st response;")
+      assert_equals(client1.getResponseHeader(state_header), id1, "1st response didn't come from server;")
+    })
+
+    setTimeout(function() {
+      req(url, id2, t, function(client2) {
+        t.step(function() {
+          if (should_have_same_body) {
+            assert_equals(client1.response, client2.response, "response bodies were different;")
+//            var res_id2 = client2.getResponseHeader(state_header)
+//            assert_not_equals(res_id2, id1, "2nd response doesn't appear to have updated cached headers;")
+//            assert_not_equals(res_id2, null, "2nd response didn't expose request identifier;")
+//            assert_equals(res_id2, id2, "2nd response is associated with a different request (!);")
+          } else {
+            assert_not_equals(client1.response, client2.response, "response bodies were the same;")
+          }
+          done(client1, client2)
+        })
+        console.log('finished checking ' + id1)
+        t.done()
+      })
+    }, 5000)
+  })
+}
+
+async_test(function(t) {
+  two_reqs('1', '2', true, t, function(client1, client2) {
+    assert_equals(client1.response, client2.response, "responses were different;")
+    assert_equals(client1.getResponseHeader("A"), null, "'A' header exposed without permission;")
+  })
+}, "A 304 response with no CORS headers inherits from the stored response")
+
+async_test(function(t) {
+  two_reqs('3', '4', true, t, function(client1, client2) {
+    assert_equals(client2.getResponseHeader("A"), "4", "304 didn't expose 'A' header, even though allowed;")
+    assert_equals(client2.getResponseHeader("B"), "4", "304 didn't expose 'B' header even though allowed;")
+  })
+}, "A 304 can expand Access-Control-Expose-Headers")
+
+async_test(function(t) {
+  two_reqs('5', '6', true, t, function(client1, client2) {
+    assert_equals(client2.getResponseHeader("B"), null, "2nd 304 exposed 'B' header;")
+  })
+}, "A 304 can contract Access-Control-Expose-Headers")
+
+async_test(function(t) {
+  two_reqs('7', '8', false, t, function(client1, client2) {
+    if (! client2.error_caught) {
```

Comment:
Not sure what I was thinking here (likely a leftover from old code), but fixed now.